File: leet_code/poly_node_tree.rb
class PolyTreeNode
  def initialize(value)
    @val = value
    @parent = nil
    @children = []
  end

  def value
    @val
  end

  def parent
    @parent
  end

  def children
    @children
  end

  def parent=(node)
    if !node
      @parent = nil
      return
    end
    return if self.parent == node
    @parent.children.delete(self) if @parent
    @parent = node
    @parent.children << self unless @parent.nil?
  end

  def add_child(child)
    child.parent = self
  end

  def remove_child(child)
    raise "no child node" if child && !@children.include?(child)
    child.parent = nil
  end
end

File: leet_code/poly_node_tree.js
class PolyTreeNode {
  constructor(value) {
    this.value = value;
    this.parent = null;
    this.children = [];
  }

  value(){
    return this.value;
  }

  parent() {
    return this.parent;
  }

  children() {
    return this.children;
  }

  setParent(node) {
    if (!node) {
      this.parent = null;
      return;
    } else if (self.parent === node) {
      return;
    }

    if (this.parent) {
      for (let i = 0; i < this.parent.children.length; i++) {
        if (this.parent.children[i] === node) {
          this.parent.children.splice(i,1);
        }
      }
    }
    this.parent = node;
    if (this.parent !== null) this.parent.children.push(this);
  }

  addChild(child) {
    child.parent = this;
  }

  childIncluded(child) {
    for (let i = 0; i < this.children.length; i++) {
      if (this.children[i] === child) return true;
    }
    return false;
  }

  removeChild(child) {
    if (child && this.childIncluded(child)) {
      throw "no child node";
    } else {
      child.parent = nil;
    }
  }
}
